Add discrete onMediaUploadPaused handler to media uploads when offline

derekblank opened the pull request:
Augments the work in https://github.com/wordpress-mobile/WordPress-Android/pull/19878 to pause media upload when internet connection is unavailable.

 This PR introduces a discrete `onMediaUploadPaused` handler to separate concerns and avoid invoking `mediaFileUploadPaused` from within `onMediaUploadFailed`. 


## To Test:

<!-- Test instructions per dependency update: https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/docs/test_instructions_per_dependency_update.md -->

-----

## Regression Notes

1. Potential unintended areas of impact

2. What I did to test those areas of impact (or what existing automated tests I relied on)

3. What automated tests I added (or what prevented me from doing so)


-----

## PR Submission Checklist:

- [ ] I have completed the Regression Notes.
- [ ] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

-----

## UI Changes Testing Checklist:

- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] Talkback.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] Large and small screen sizes. (Tablet and smaller phones)
- [ ] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

derekblank committed:
Add discrete onMediaUploadPaused handler to media uploads when offline

derekblank commented on libs/editor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergEditorFragment.java:
```diff
@@ -1488,14 +1488,14 @@ public void onMediaUploadProgress(final String localMediaId, final float progres
+    @Override
+    public void onMediaUploadPaused(final String localMediaId) {
+        getGutenbergContainerFragment().mediaFileUploadPaused(Integer.valueOf(localMediaId));
         mFailedMediaIds.add(localMediaId);
```
I don't really wish to add these localMediaIds to `mFailedMediaIds`, but it is necessary to automatically restart media uploads:

https://github.com/wordpress-mobile/WordPress-Android/blob/234db297728a506ee8e9c318d42c9cd9e7e941b3/libs/editor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergEditorFragment.java#L1591-L1596

Semantically, this should probably be called `mFailedOrPausedMediaIds` (including related artifacts 
like `hasFailedMediaUploads` -> `hasFailedOrPausedMediaUploads`). The concept of failed uploads could be rethought of entirely with the introduction of `MEDIA_UPLOAD_STATE_PAUSED`. That feels well outside the scope of this change, as it has quite a large surface area. If there is a more elegant way to restart media uploads without adding these to `mFailedMediaIds`, I am open to it.

dcalhoun replied:
I agree a refactor is likely beneficial but unnecessary at this time. 

That said, I do not believe overloading `mFailedMediaIds` to include "paused" media is all that cryptic or misleading, particularly if we only consider media to be "paused" if it also contains an error, as I noted in a separate comment.

dcalhoun commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:
```diff
@@ -3678,6 +3678,11 @@ public void onMediaUploaded(OnMediaUploaded event) {
             return;
         }
 
+        if (!NetworkUtils.isNetworkAvailable(this)) {
```
I worry that if we do not check for an error as well, we might erroneously mark successfully uploaded media as paused. WDYT?

dcalhoun commented on WordPress/src/main/java/org/wordpress/android/ui/posts/editor/media/EditorMedia.kt:
```diff
@@ -298,6 +298,10 @@ class EditorMedia @Inject constructor(
         listener.onMediaUploadFailed(media.id.toString(), error.type.name)
     }
 
+    fun onMediaUploadPaused(listener: EditorMediaUploadListener, media: MediaModel) = launch {
+        listener.onMediaUploadPaused(media.id.toString())
+    }
```
I imagine we should still track an event of `EDITOR_UPLOAD_MEDIA_FAILED` for paused media, or implement a new event. WDYT?

dcalhoun commented on libs/editor/src/main/java/org/wordpress/android/editor/EditorMediaUploadListener.java:
```diff
@@ -9,4 +9,5 @@ public interface EditorMediaUploadListener {
     void onMediaUploadProgress(String localId, float progress);
     void onMediaUploadFailed(String localId, String errorType);
     void onGalleryMediaUploadSucceeded(long galleryId, long remoteId, int remaining);
+    void onMediaUploadPaused(String toString);
```
This is an odd parameter name, presumably a typo. 
```suggestion
    void onMediaUploadPaused(String localId);
```

dcalhoun commented on libs/editor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java:
```diff
@@ -1431,6 +1431,10 @@ public void onMediaUploadFailed(final String localMediaId, final String errorTyp
         mUploadingMediaProgressMax.remove(localMediaId);
     }
 
+    public void onMediaUploadPaused(final String localMediaId) {
+        mUploadingMediaProgressMax.remove(localMediaId);
+    }
```
I like avoiding a need to pass the error type and entangle the paused and error methods, but I worry about modify Aztec. We likely need to test Aztec to avoid regressions. 

Additionally, I note this method lacks an `@Override` annotation that surround methods possess. We should likely follow the existing code pattern. 

Lastly, how confident are we that we only need to remove the `localMediaId`? Do we need the guard clause or other logic from the sibling failed method?

derekblank replied:
Adding `onMediaUploadPaused` to Aztec was done quickly to get around some build errors from Android Studio. I also am not keen on modifying Aztec. I'll do some further testing to see if this is actually needed.

dcalhoun replied:
My understanding is that it is necessary, as both Aztec and Gutenberg share a common interface to which they must adhere. For Aztec, the latest implementation of this merely passes the invocation onto the existing failed method. So, the regression risk should be fairly low. It is likely still worth testing the Aztec editor via editing classic posts in the app. 

https://github.com/wordpress-mobile/WordPress-Android/blob/b97dbeccb556b565249cfe6d2ab229ba9479d33a/libs/editor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java#L1434-L1438

dcalhoun commented on libs/editor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergEditorFragment.java:
```diff
@@ -1488,14 +1488,14 @@ public void onMediaUploadProgress(final String localMediaId, final float progres
 
     @Override
     public void onMediaUploadFailed(final String localMediaId, String errorType) {
```
With this approach, we no longer require the `errorType` parameter and should remove it. 
```suggestion
    public void onMediaUploadFailed(final String localMediaId) {
```

dcalhoun reviewed (commented):
I like your proposed changes. I plan to checkout the branch locally and implement some of my suggested changes. I am still interested in your responses to my suggestions, though.

dcalhoun committed:
refactor: Fix parameter name typo

dcalhoun committed:
refactor: Annotate override of `onMediaUploadPaused` method

Ensure the implementation adheres to the relevant interface.

dcalhoun committed:
refactor: Remove unused `onMediaUploadFailed` error type parameter

With the addition of a discrete paused method, we no longer utilize this
parameter in the failed method.

dcalhoun committed:
feat: Aztec `onMediaUploadPaused` passes onto `onMediaUploadFailed`

We do not plan to add paused media support to the Aztec editor.
Therefore, it likely makes more sense for the `onMediaUploadPaused`
implementation to merely invoke the existing `onMediaUploadFailed`
method.

dcalhoun committed:
feat: Only consider failed media uploads to be paused

Without checking for an error, there may be a possibility that
successfully uploaded media is marked as "paused."

dcalhoun committed:
feat: Track failed media upload event when pausing media uploads

Even though we communicate a nuanced paused state to the user, we likely
need to continue tracking this event as a failure, or consider adding a
discrete pause analytic event.

dcalhoun committed:
feat: When lacking connectivity, pause completed upload w/o remote URLs

Based upon existing code, we should consider completed uploads lacking
remote URLs to be failures. Therefore, we should apply the new pause
state to this context.

dcalhoun commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:
```diff
@@ -3696,7 +3701,10 @@ public void onMediaUploaded(OnMediaUploaded event) {
         } else if (event.completed) {
             // if the remote url on completed is null, we consider this upload wasn't successful
-            if (TextUtils.isEmpty(event.media.getUrl())) {
+            if (TextUtils.isEmpty(event.media.getUrl()) && !NetworkUtils.isNetworkAvailable(this)) {
+                MediaError error = new MediaError(MediaErrorType.GENERIC_ERROR);
+                mEditorMedia.onMediaUploadPaused(mEditorMediaUploadListener, event.media, error);
```
Noting that I added the paused media implementation here. Based upon the existing code, it appears that completed media should be considered failed if it possesses a null remote URL.
